src/Servers/Kestrel/Core/src/Internal/Infrastructure/Heartbeat.cs

```diff
@@ -71,17 +73,35 @@ internal void OnHeartbeat()
 
     private void TimerLoop()
     {
-        Thread.Sleep(_interval);
+        // Starting the heartbeat immediately triggers OnHeartbeat.
+        // Initial delay to avoid running heartbeat again from timer thread.
+        if (_stopWaitHandle.WaitOne(_interval))
+        {
+            return;
+        }
+
         while (!_stopped)
         {
             OnHeartbeat();
-            Thread.Sleep(_interval);
+
+            if (_stopWaitHandle.WaitOne(_interval))
+            {
+                return;
+            }
```

Comment:
I guess you could just do this, and add a check for _stopped at the top of OnHeartbeat.
```c#
    private void TimerLoop()
    {
        // Starting the heartbeat immediately triggers OnHeartbeat.
        // Initial delay to avoid running heartbeat again from timer thread.
        while (!_stopped)
        {
            if (_stopWaitHandle.WaitOne(_interval))
            {
                return;
            }

            OnHeartbeat();
      }
   }
```

Reply:
actually, why do you need `_stopped` any more?

Reply:
You're right that it's pretty redundant here. `_stopped` it's also used to suppress logging errors from the heartbeat as the server shuts down.

Better behavior is to completely stop the heartbeat before starting shutdown. Can eliminate `_stopped` once that is done.